great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -1558,11 +1565,16 @@ def list_datasources(self) -> List[dict]:
                 datasource_config = DatasourceConfig(**datasource_config)
             datasource_config.name = datasource_name
 
-            masked_config: dict = (
-                self._serialize_substitute_and_sanitize_datasource_config(
+            serialized_config: dict = (
+                self._serialize_datasource_config(
                     serializer, datasource_config
                 )
             )
+            masked_config = self._substitute_and_sanitize_datasource_config(serialized_config)
+            datasources.append(masked_config)
+
+        for datasource_name, datasource_config in self.config.fluent_datasources.items():
+            masked_config = self._substitute_and_sanitize_datasource_config(datasource_config)
```

Comment:
We do config substitution differently for fluent datasources.
For BDS we eagerly do substitute configs but for FDS we do it lazily, so I don't know that `_substitute_and_sanitize_datasource_config` makes sense here.

Reply:
Ah okay I may have led @jshaikGX astray here then.

Given how substitution is done for fluent datasources, are we ever concerned about credential leakage? I was going to recommend a test that ensures credential obfuscation but if it isn't relevant lmk

Reply:
Because of the lazy loading there are no credentials to actually leak. So it's inherintly safer. 
You have to call `.get_config_value()` on the `ConfigStr` type to do the substitution.

Reply:
@Kilo59 Let me know if this isnt relevant

Reply:
Also, if you remove like 1577 I think your type checking problem will go away.